File: sdk-dotnet/tests/HealthFlow.Hfcx.Sdk.Tests/CrossSdkFixtures.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.IO;
using System.Security.Cryptography;

namespace HealthFlow.Hfcx.Sdk.Tests;

/// <summary>
/// Loaders for the cross-SDK round-trip fixtures shared with the Java and
/// Python SDKs. Every SDK reads <c>private-key.pem</c> /
/// <c>public-key.pem</c> / <c>plaintext.json</c> from the same byte-for-byte
/// fixtures so any drift in JWE wire format shows up immediately as a test
/// failure.
/// </summary>
internal static class CrossSdkFixtures
{
    public static string FixtureDir
    {
        get
        {
            var dir = Path.Combine(AppContext.BaseDirectory, "fixtures", "cross-sdk");
            if (!Directory.Exists(dir))
            {
                throw new DirectoryNotFoundException(
                    $"cross-sdk fixtures missing at {dir}; check csproj Copy rule");
            }

            return dir;
        }
    }

    public static RSA LoadPrivateKey()
    {
        var rsa = RSA.Create();
        rsa.ImportFromPem(File.ReadAllText(Path.Combine(FixtureDir, "private-key.pem")));
        return rsa;
    }

    public static RSA LoadPublicKey()
    {
        var rsa = RSA.Create();
        rsa.ImportFromPem(File.ReadAllText(Path.Combine(FixtureDir, "public-key.pem")));
        return rsa;
    }

    public static string LoadPlaintext()
        => File.ReadAllText(Path.Combine(FixtureDir, "plaintext.json"));

    public static string LoadJavaProducedJwe()
        => File.ReadAllText(Path.Combine(FixtureDir, "java-produced.jwe")).Trim();

    public static string LoadPythonProducedJwe()
        => File.ReadAllText(Path.Combine(FixtureDir, "python-produced.jwe")).Trim();

    public static string LoadJavaScriptProducedJwe()
        => File.ReadAllText(Path.Combine(FixtureDir, "javascript-produced.jwe")).Trim();
}

File: sdk-javascript/tests/unit/crossSdkFixtures.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { type KeyLike, importPKCS8, importSPKI } from 'jose';

import { JWE_ALG } from '../../src/crypto/JweAlgorithms.js';

const here = dirname(fileURLToPath(import.meta.url));
export const FIXTURE_DIR = join(here, '..', 'fixtures', 'cross-sdk');

export async function loadPrivateKey(): Promise<KeyLike> {
  const pem = readFileSync(join(FIXTURE_DIR, 'private-key.pem'), 'utf8');
  return importPKCS8(pem, JWE_ALG);
}

export async function loadPublicKey(): Promise<KeyLike> {
  const pem = readFileSync(join(FIXTURE_DIR, 'public-key.pem'), 'utf8');
  return importSPKI(pem, JWE_ALG);
}

export function loadPlaintext(): string {
  return readFileSync(join(FIXTURE_DIR, 'plaintext.json'), 'utf8');
}

export function loadJwe(producer: 'java' | 'python' | 'dotnet' | 'javascript'): string {
  return readFileSync(join(FIXTURE_DIR, `${producer}-produced.jwe`), 'utf8').trim();
}
